File: Typescript/projects/linql.core/src/lib/LinqlExpression.ts
export abstract class LinqlExpression
{
    abstract "$type": string;

    public Next: LinqlExpression | undefined;

    public GetLastExpressionInNextChain(): LinqlExpression
    {
        if (!this.Next)
        {
            return this;
        }
        return this.Next.GetLastExpressionInNextChain();
    }

    public abstract Clone(): this;
}

File: Python/linql-core/src/linql_core/LinqlExpression.py
from typing_extensions import Self
import abc

class LinqlExpression(abc.ABC):
    
    type: str = None
    Next: Self = None
    
    def GetLastExpressionInNextChain(self) -> Self:
        if self.Next == None:
            return self
        return self.Next.GetLastExpressionInNextChain()

    @abc.abstractmethod
    def Clone(self) -> Self:
        pass
    
    @abc.abstractmethod
    def toSerializable(self) -> object:
        pass

    def _CreateSerializableType(self) -> dict:
        serializeable = { "$type": self.type }
        return serializeable
    
    def _SerializeNext(self, existingObject: dict) -> dict:
        if hasattr(self, "Next") and self.Next != None:
            existingObject["Next"] = self.Next.toSerializable()
        return existingObject